evm-precompiles/erc20/src/lib.rs

```diff
@@ -358,57 +357,53 @@ where
+				Some(from).into(),
+				crml_generic_asset::Call::<Runtime>::transfer { asset_id, to, amount },
+				gasometer,
+			)?;
+		} else {
+			// caller not from, check if caller is approved
 			gasometer.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
-
 			let current_approved_amount: Balance =
 				crml_token_approvals::Module::<Runtime>::erc20_approvals((&from.clone(), &asset_id), &context.caller);
-			if current_approved_amount >= amount {
-				let new_approved_amount: Balance = current_approved_amount - amount;
-				gasometer.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
-
-				if new_approved_amount == Balance::zero() {
-					// New balance is 0, remove approval
-					RuntimeHelper::<Runtime>::try_dispatch(
-						None.into(),
-						crml_token_approvals::Call::<Runtime>::erc20_remove_approval {
-							caller: from.clone(),
-							asset_id,
-							spender: context.caller,
-						},
-						gasometer,
-					)?;
-				} else {
-					// New balance has changed, update approval to represent difference
-					RuntimeHelper::<Runtime>::try_dispatch(
-						None.into(),
-						crml_token_approvals::Call::<Runtime>::erc20_approval {
-							caller: from.clone(),
-							spender: context.caller,
-							asset_id,
-							amount: new_approved_amount,
-						},
-						gasometer,
-					)?;
-				}
-
-				let from: Runtime::AccountId = Runtime::AddressMapping::into_account_id(from.clone());
-				let to: Runtime::AccountId = Runtime::AddressMapping::into_account_id(to);
-				gasometer.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
+			let new_approved_amount: Balance = current_approved_amount
+				.checked_sub(amount)
+				.ok_or(error("Caller not approved for amount").into())?;
 
+			gasometer.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
+			if new_approved_amount.is_zero() {
+				// New balance is 0, remove approval
 				RuntimeHelper::<Runtime>::try_dispatch(
-					Some(from).into(),
-					crml_generic_asset::Call::<Runtime>::transfer { asset_id, to, amount },
+					None.into(),
+					crml_token_approvals::Call::<Runtime>::erc20_remove_approval {
+						caller: from.clone(),
+						asset_id,
+						spender: context.caller,
+					},
 					gasometer,
 				)?;
 			} else {
-				return Err(error("caller not approved").into());
+				// New balance has changed, update approval to represent difference
+				RuntimeHelper::<Runtime>::try_dispatch(
+					None.into(),
+					crml_token_approvals::Call::<Runtime>::erc20_approval {
+						caller: from.clone(),
+						spender: context.caller,
+						asset_id,
+						amount: new_approved_amount,
+					},
+					gasometer,
+				)?;
 			}
-		} else {
+
 			let from: Runtime::AccountId = Runtime::AddressMapping::into_account_id(from.clone());
 			let to: Runtime::AccountId = Runtime::AddressMapping::into_account_id(to);
-			// Dispatch call (if enough gas).
+			gasometer.record_cost(RuntimeHelper::<Runtime>::db_write_gas_cost())?;
```

Comment:
`try_dispatch` does its own gas metering so don't need to record the write cost separately